src/storage/src/client/controller.rs

```diff
@@ -495,71 +495,85 @@ where
-                let remap_write = self
-                    .persist_client
-                    .open_writer::<(), PartitionId, T, MzOffset>(metadata.remap_shard)
-                    .await
-                    .unwrap();
-                for t in remap_write.upper().elements() {
-                    resume_upper.insert(t.clone());
+                    // Check if this ingestion is using any operators that are stateful AND are not
+                    // storing their state in persist shards. This whole section should be eventually
+                    // removed as we make each operator durably record its state in persist shards.
+                    let resume_upper = match ingestion.desc.envelope {
+                        // We can only resume with the None envelope right now which is stateless
+                        SourceEnvelope::None(_) => resume_upper,
+                        // Otherwise re-ingest everything
+                        _ => Antichain::from_elem(T::minimum()),
+                    };
+
+                    let augmented_ingestion = IngestSourceCommand {
+                        id,
+                        description: IngestionDescription {
+                            source_imports,
+                            storage_metadata: self.collection(id)?.collection_metadata.clone(),
+                            // The rest of the fields are identical
+                            desc: ingestion.desc,
+                            typ: description.desc.typ().clone(),
+                        },
+                        resume_upper,
+                    };
+
+                    // Provision a storage host for the ingestion.
+                    let client = self
+                        .hosts
+                        .provision(id, description.remote_addr.clone())
+                        .await?;
+
+                    client.send(StorageCommand::IngestSources(vec![augmented_ingestion]));
                 }
-                let data_write = self
-                    .persist_client
-                    .open_writer::<SourceData, (), T, Diff>(metadata.data_shard)
-                    .await
-                    .unwrap();
-                for t in data_write.upper().elements() {
-                    resume_upper.insert(t.clone());
+                None => {
+                    // If this collection doesn't have an ingestion associated with it will be
+                    // managed by append calls to the storage controller. So immediately update our
+                    // write frontier to be the current global upper to be prepared for any future
+                    // appends.
+                    let mut change_batch = ChangeBatch::new();
+                    change_batch.extend(upper.into_iter().map(|t| (t, 1)));
+                    change_batch.update(T::minimum(), -1);
+                    self.update_write_frontiers(&[(id, change_batch)]).await?;
```

Comment:
We also execute this part for recorded view collections, which are not advanced through `append` calls, but directly using `update_write_frontiers` calls in the compute controller: https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/controller.rs#L697-L709

Advancing `since`s for those collections currently works correctly, which means it will probably stop working with this change. I guess an easy fix would be moving the frontier updates for recorded views to use `append` as well.